### DISCUSSION TITLE
Is CylcloneDX able to describe a hardware software stack?

### DISCUSSION THREAD

**Discussion by 3BK** [CycloneDX/specification#243] [Q&A]
Is CylcloneDX able to describe a hardware software stack?

How do we track components such as database servers and databases muchless containers and pods?

Looking at the list of components in CycloneDX, I see: Application, Framework, Operating System, Firmware, Library, Container, Device and file.

then comparing to a simplified hardware stoftware stack, I see: Applications, Middleware, database, operating system, virtual machine, server, storage.    

Sorting the components by hardware stack layer - we have:

- application tier:
   - components:
      - Application, Framwork, Library and file 
      
- Middware tier: 
  - can we reuse components from the application tier? e.g. application, framework, library and file
  - How do we identify secure web gateways interfaces?
  - how do we identify database drivers?
  - Are we missing a middlware component?
  - Question does the container tier fall under middleware or is it separate? 
  
- Database tier: 
   - can we reuse components from the middleware tier?
   - How do we show that the application is dependent on a database server?
   - Are we missing a database server component?
   - How do we show that the application is dependent on a database?
   - Are we missing a database component?
   - How do we show that the dabase is dependent on a group of files?
   - Are we missing a database file group component?

- OS tier:
   -  can we reuse components from the database tier?
   -  what components are unique to the OS interrupt? Clock? driver?
   -  should we add them to the list of components?

 - virtual machine tier:
    -  what components are unique to the vm tier?  vTPM?  vClock, VHBA, vNIC?  
    -  should we add them to the list of components?

 - server tier:
   -  what components are unique to the server tier?  device?  Power?
   -  should we add them to the list of components?

 - storage tier:
    -  what components are unique to the storage tier?  device?  Power?
    -  should we add them to the list of components?

**Comment by 3BK** [comment]
I think this is a defect - not a question. CycloneDX claims to be able to describe a full stack - muchless services.

**Reply by jkowalleck** [reply]
please wait for an answer, maybe you just missed how this capabilities are done.

**Comment by stevespringett** [comment]
For hardware, please take a look at https://cyclonedx.github.io/cyclonedx-property-taxonomy/cdx/device and the example that we provide in GitHub. There are also folks from large manufactures that are producing HBOM or have working POCs with CycloneDX HBOMs in the Slack workspace.

Component types are abstract concepts. Please do not think of them as concrete. It would be impossible to come up with a complete taxonomy of different component types. That is not the goal. The goal was to achieve a certain amount of separation of concerns.

Middleware is just another application. If you want to use properties to expand upon that, you can do so. You can also use services to describe the services that the middleware provides or depends on. Same thing for database.

There are a few new component types in v1.5 including platform, device driver, data, and machine learning model. 

Storage is a bit odd. We can describe network storage such as S3 or NAS using services. We can also describe the physical storage device. But currently, we cannot describe the volume. That capability is coming with v1.5, but will initially be limited to formulation support, not inventory support.

**Reply by 3BK** [reply]
I'm having trouble seeing how we can describe a on prem BOM much less SaaS using CycloneDX

**Comment by 3BK** [comment]
Hi Steve.
Thanks for your response.  I'm still not seeing the database layer.  I will embed a dotty graph to illustrate my point.

**Reply by jkowalleck** [reply]
what do you understand by "database layer"? 
is it file-based like SQLite, or is it a DBMS like MySQL, or is it a service like graphQL? 
Hope this graph will make it clear, because the answer depends ...

**Reply by 3BK** [reply]
Please see the **standard n-tier model**.

For example this patent from 2001 [Method and apparatus for efficient SQL processing in an n-tier architecture](https://patents.google.com/patent/US7580971B1/en)

A true database layer.  
 - Not a tightly coupled datastore like sqlite or mysql.  

A loosely coupled database tier which is fault tolerant.

**Reply by 3BK** [reply]
A real database tier addresses sensor, actuator, bus, and node failure.

**Comment by 3BK** [comment]
digraph G {


  srv1_App1 -> srv1;
  srv1_App1 -> srv2;
  srv1_App1 -> srv3;
  srv1_App1 -> srv4;


  srv1_App1 -> srv2_App1;

  srv1_App1 -> srv1_sw1;
  srv1_App1 -> srv2_data1;
  
  srv2_app1_sw1 -> srv1_library;
  srv2_app1_sw1 -> srv1_OS;

  srv2_App1 -> srv2_app1_sw1;
  srv2_App1 -> srv2_app1_data1;

  srv2_app1_sw1 -> srv2_library;
  srv2_app1_sw1 -> srv2_OS;

  srv2_app1_data1     ->  srv2_app1_data1_fs1;

  srv2_app1_data1_fs1 -> srv2_file1;
  srv2_app1_data1_fs1 -> srv2_file2;
  srv2_app1_data1_fs1 -> srv2_file3;
  srv2_app1_data1_fs1 -> srv2_file4;

  srv2_file1 -> srv3;
  srv2_file2 -> srv4;
  srv2_file3 -> srv3;
  srv2_file4 -> srv4;
  
  srv3 -> srv3_dsk1;
  srv3 -> srv3_dsk2;
  srv3 -> srv3_dsk3;
  srv3 -> srv3_dsk4;

  srv4 -> srv4_dsk1;
  srv4 -> srv4_dsk2;
  srv4 -> srv4_dsk3;
  srv4 -> srv4_dsk4;

  srv3_dsk1 -> srv3_dsk1_fw1;
  srv3_dsk1 -> srv3_dsk1_data1;
  srv3_dsk2 -> srv3_dsk2_fw1;
  srv3_dsk2 -> srv3_dsk2_data1;
  srv3_dsk3 -> srv3_dsk3_fw1;
  srv3_dsk3 -> srv3_dsk3_data1;
  srv3_dsk4 -> srv3_dsk4_fw1;
  srv3_dsk4 -> srv3_dsk4_data1;

  srv4_dsk1 -> srv4_dsk1_fw1;
  srv4_dsk1 -> srv4_dsk1_data1;
  srv4_dsk2 -> srv4_dsk2_fw1;
  srv4_dsk2 -> srv4_dsk2_data1;
  srv4_dsk3 -> srv4_dsk3_fw1;
  srv4_dsk3 -> srv4_dsk3_data1;
  srv4_dsk4 -> srv4_dsk4_fw1;
  srv4_dsk4 -> srv4_dsk4_data1;
  
}

**Comment by 3BK** [comment]
I mean a database tier such as DB2 HADR, Oracle RAC, SQL Server ALWAYS ON, or Postgresql HA consisting of at least 2 voted nodes.  

Maybe 2oo3.
Maybe 2oo4 + 1.

**Reply by stevespringett** [reply]
Ok. So you need multiple OBOMs. Simple. Each OBOM would reference cooresponding SBOM and HBOM for the assets. That allows you to describe physical assets with hardware, add-ons, operating systems, firmware, applications, etc.

**Comment by 3BK** [comment]
I don't mean a tightly coupled datastore like sqlite or mysql .

I mean a database server.  One which earns revenue.

**Comment by 3BK** [comment]
![graphviz (1)](https://github.com/CycloneDX/specification/assets/40867233/ad2769f3-fd21-4391-b493-078fc61836dc)

**Reply by 3BK** [reply]
Please note this is a simplified diagram

**Reply by stevespringett** [reply]
That can easily be represented. What's the question?

**Comment by 3BK** [comment]
Note that the stack needs software and data packages as per CIS and STIG.

**Comment by 3BK** [comment]
Unless you want to call data packages software/firmware which can't be executed.  I'm OK with that but it will blow some peoples minds.

**Reply by stevespringett** [reply]
As I mentioned earlier. Theres a new component type in v1.5 called `data`.

**Comment by 3BK** [comment]
To elaborate, if you call data software which can't be executed, then what do you a call a register.  Something which streams Software ?  Then what do you call a demux, something which streams multiple software elements?

**Reply by stevespringett** [reply]
Why are there so many threads? If you have a question, and related questions, please use threading. It's impossible to keep track of this conversation.

**Comment by 3BK** [comment]
I won't even ask about how we specify the dependency on the HVAC much less single mode fibre.

**Reply by stevespringett** [reply]
Same way as you would specify dependencies on anything else. You do realize that the CycloneDX data model was inspired by things in the physical supply chain. Specifically pharmaceuticals, but can easily be applied to automotive, etc. In fact, describing the guidance system on an F-35 is a real use case for us. So please do not make assumptions.

**Comment by 3BK** [comment]
Oops.  The temperature is 90+ Deg C - BACNET is down.  We just fried the equipment due to Electrostatic discharge(ESD) and floating ground. The battery pile caught on fire. And, the pipes burst.

**Reply by stevespringett** [reply]
Is this suppose to be a question? I don't see how operational constraints belong in a BOM. They do not. However, you can use `risk assessment` in v1.5 to specify an external location of where those risks are documented at. Or you could use the existing `documentation` external ref to specify a location where those constraints are documented.

**Comment by 3BK** [comment]
The GPS-R attena can't see any GPS-T satellites.  Someone took the Cesium clock - unable to authenticate to the domain.  Access control system is DOA - unable to open the door.  Fire supression is offline - looking for a stick, some wieners, and some marshmellows.   Because they weren't documented in CycloneDX

**Reply by stevespringett** [reply]
What is the question?

**Comment by 3BK** [comment]
Where are the "45 TB"  8 track tapes? Oops, no one put the firesafe in cycloneDX. I guess we didn't document the HW/SW stack. ;)

**Reply by stevespringett** [reply]
What is the question?

**Comment by 3BK** [comment]
Let's not even discuss what's included in SaaS. ;)

**Reply by stevespringett** [reply]
SaaSBOM is a real thing and multiple open source and commercial solutions are available.

**Comment by 3BK** [comment]
[like]  Brent Kimberley reacted to your message:
________________________________

**Comment by 3BK** [comment]
Perfect!
________________________________
